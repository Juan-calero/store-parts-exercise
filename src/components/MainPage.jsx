import React, { useState, useEffect } from "react";
import Product from "./Product";
import {
	mainSection,
	mainForm,
	productSection,
	priceOrder,
} from "./mainpage.module.scss";

const MainPage = () => {
	const [productsData, setProductsData] = useState([]);
	const [searchValue, setSearchValue] = useState("");
	const [types, setTypes] = useState([]);
	const [typeValue, setTypeValue] = useState("Type");
	const [isAscendingPrice, setIsAscendingPrice] = useState(true);

	useEffect(() => {
		const callAPI = async () => {
			const resp = await fetch("http://localhost:8081/store/parts");
			if (resp.status >= 400) {
				throw new Error("Oh uh, something went wrong");
			}
			const body = await resp.json();
			body.sort((a, b) =>
				+a.price.slice(0, -1) > +b.price.slice(0, -1) ? 1 : -1
			);
			setProductsData(body);
		};
		callAPI();
	}, []);

	useEffect(() => {
		const callAPI = async () => {
			const resp = await fetch("http://localhost:8081/store/part-types");
			if (resp.status >= 400) {
				throw new Error("Oh uh, something went wrong");
			}
			const body = await resp.json();
			setTypes(body);
		};
		callAPI();
	}, []);

	const handlePriceOrder = (e) => {
		e.preventDefault();
		setProductsData(productsData.reverse());
		setIsAscendingPrice(!isAscendingPrice);
	};

	const filter = (product) => {
		const checkType = typeValue === "Type" || product.type === typeValue;
		const checkSearch =
			!searchValue || product.name.toLowerCase().includes(searchValue);
		return checkSearch && checkType && product;
	};

	const mapProducts = productsData.filter(filter).map((product) => {
		return <Product key={product.name} {...product} />;
	});
	const mapTypes = types.map((type, index) => {
		return (
			<option key={index} value={type}>
				{type}
			</option>
		);
	});

	return (
		<main className={mainSection}>
			<form className={mainForm}>
				<input
					aria-label="Search"
					type="search"
					placeholder="Search..."
					value={searchValue}
					onChange={(e) => setSearchValue(e.target.value)}
				/>
				<select
					aria-label="Type"
					onChange={(e) => setTypeValue(e.target.value)}
				>
					<option value="Type">Type</option>
					{mapTypes}
				</select>
				<input
					aria-label="Price Order"
					className={priceOrder}
					type="button"
					value={`${isAscendingPrice ? "Ascending" : "Descending"} Price`}
					onClick={(e) => handlePriceOrder(e)}
				/>
			</form>
			<section className={productSection}>
				<ul>{mapProducts}</ul>
			</section>
		</main>
	);
};

export default MainPage;
