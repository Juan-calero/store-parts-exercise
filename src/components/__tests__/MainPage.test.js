import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import Enzyme, { shallow } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import MainPage from "../MainPage";

Enzyme.configure({ adapter: new Adapter() });

afterEach(() => {
	cleanup();
});

//SNAPSHOT
test("should match snapshot", () => {
	render(<MainPage />);
	const tree = renderer.create(<MainPage />).toJSON();
	expect(tree).toMatchSnapshot();
});

//INPUT FIELDS
test("Search field is empty on render", () => {
	const { getByLabelText } = render(<MainPage />);
	expect(getByLabelText(/Search/).textContent).toBe("");
});

test("button should render in Ascending Price", () => {
	const wrapper = shallow(<MainPage />);
	const text = wrapper.find(".priceOrder");
	expect(text.get(0).props.value).toContain("Ascending");
});

test("button should become Descending Price after click", () => {
	const wrapper = shallow(<MainPage />);
	const button = wrapper.find(".priceOrder");
	button.simulate("click", {
		preventDefault: () => {},
	});
	const text = wrapper.find(".priceOrder");
	expect(text.get(0).props.value).toContain("Descending");
});

test("Select field should start with TYPE option", () => {
	const { getByText } = render(<MainPage />);
	expect(getByText(/Type/).textContent).toBe("Type");
});

//FILTERING PROCESS
describe("Filtering Should work", () => {
	const filter = (product, typeValue, searchValue) => {
		const checkType = typeValue === "Type" || product.type === typeValue;
		const checkSearch =
			!searchValue || product.name.toLowerCase().includes(searchValue);
		return checkSearch && checkType && product;
	};
	test("Filter Mouse with type:Type and search:mou. Should be Truthy.", () => {
		const product = {
			name: "Mouse 2",
			price: "35.99$",
			type: "Mouse",
		};
		expect(filter(product, "Type", "mou")).toBeTruthy();
	});
	test("Filter Keyboard with type:Type and search:mou. Should be Falsy.", () => {
		const product = {
			name: "Keyboard 3",
			price: "83.00$",
			type: "Keyboard",
		};
		expect(filter(product, "Type", "mou")).toBeFalsy();
	});
	test("Filter Monitor with type:Monitor and search:itor. Should be Truthy", () => {
		const product = {
			name: "Monitor 1",
			price: "531.00$",
			type: "Monitor",
		};
		expect(filter(product, "Monitor", "itor")).toBeTruthy();
	});
});
