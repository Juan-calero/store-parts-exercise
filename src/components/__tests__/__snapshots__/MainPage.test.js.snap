// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<main
  className="mainSection"
>
  <form
    className="mainForm"
  >
    <input
      aria-label="Search"
      onChange={[Function]}
      placeholder="Search..."
      type="search"
      value=""
    />
    <select
      aria-label="Type"
      onChange={[Function]}
    >
      <option
        value="Type"
      >
        Type
      </option>
    </select>
    <input
      aria-label="Price Order"
      className="priceOrder"
      onClick={[Function]}
      type="button"
      value="Ascending Price"
    />
  </form>
  <section
    className="productSection"
  >
    <ul />
  </section>
</main>
`;
